feat: Fail on change

olivier-lacroix opened the pull request:
#### :zap: Summary

This PR introduces a new `fail_on_changes` configuration option. The intent is that in certain situations (e.g. CI), any file modification indicates an issue. See https://github.com/evilmartians/lefthook/discussions/580 for instance

If `fail_on_changes` is

- `never`, lefthook will behave identically to what it does today
- `always`, lefthook will throw an error if it modifies any files in the repository
- `ci`, lefthook will behave identically to what it does today, unless a `CI` environment variable is present (whatever its value), in which case it will throw an error if it modifies any files in the repository.

<!-- Brief description of what was done -->

#### :ballot_box_with_check: Checklist

- [X] Check locally
- [x] Add tests
- [x] Add documentation

olivier-lacroix commented on internal/git/repository.go:
```diff
@@ -158,7 +158,7 @@ func (r *Repository) Setup() {
 
 	r.statusShortOnce = sync.OnceValues(func() ([]string, error) {
-		return r.Git.CmdLines(cmdStatusShort)
+		return r.Git.CmdLinesNoTrim(cmdStatusShort)
```
Unrelated to this PR, note that `Git.CmdLines` trimming leading space would create issues when the first file returned by `git status --short` is unstaged, as the leading space is, in that case, meaningful.

olivier-lacroix commented:
@mrexox no worries! Thanks for your message. I will be offline for a week or so, so take your time :)

mrexox commented on internal/config/config.go:
```diff
@@ -54,6 +54,8 @@ type Config struct {
 
 	Templates map[string]string `json:"templates,omitempty" jsonschema:"description=Custom templates for replacements in run commands." mapstructure:"templates,omitempty"`
 
+	FailOnChanges *bool `json:"fail_on_changes,omitempty" jsonschema:"description=Exit with non-zero status if any files were modified." koanf:"fail_on_changes" mapstructure:"fail_on_changes,omitempty"`
```
Oh I see, this way checking changes makes sense.

I don't have a use-case in mind about job-level option, it just feels similar to `stage_fixed` by scope. Like, some jobs edit files and user may want to make sure they do not change files. But that's only an assumption.

Do you think people may have different flows in CI and on local machines? Like, locally I want pre-commit to stage the changes, but in the CI I need to `exit 1` if changes were staged? If so, WDYT about adding a **CLI option**?

So, I suggest the following:
- Move the option to the hook config
- Add a CLI option `--fail-on-changes`

I'd be glad to know what you think about this.

olivier-lacroix replied:
Happy with that @mrexox !

Indeed, the use case you describe is exactly what I would expect

olivier-lacroix commented:
@mrexox I have moved the configuration to the hook level, and added a CLI argument.

Note that although I think it is a sensible default, the current behaviour in this PR is somewhat of a breaking change in environments where the CI env var is set, as changes will trigger a failure.

If this is an issue, an alternative would be to make `fail_on_change` a setting that would take `never`, `always`, and `ci` as values, defaulting to `never` to replicate the current behaviour.

Any thoughts on this?

olivier-lacroix commented:
@mrexox I have updated the config so that the default behaviour remains the same.

I have kept the ability to (explicitly) rely on the presence of a CI env var to modify the behaviour. Using that env var to modify behaviour is fairly common - after all this is the main purpose of that env var. Hope this is okay 😄 .
